plugins/osdn/netid/controller/controller.go

```diff
@@ -0,0 +1,170 @@
+package controller
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/openshift/openshift-sdn/pkg/netid"
+	"github.com/openshift/openshift-sdn/plugins/osdn/netid/vnidallocator"
+)
+
+type Allocation struct {
+	vnid   vnidallocator.Interface
+	client kclient.NamespaceInterface
+	// globalNamespaces are the namespaces that have access to all pods in the cluster and vice versa.
+	globalNamespaces sets.String
+}
+
+// retryCount is the number of times to retry on a conflict when updating a namespace
+const retryCount = 2
+
+// AddOrUpdate allocates network id(VNID) for the namespace if needed
+func (c *Allocation) AddOrUpdate(ns *kapi.Namespace) error {
+	tx := &tx{}
+	defer tx.Rollback()
+
+	var id uint
+	userRequestedVnid := false
+	value, err := netid.GetWantsVNID(ns)
+	if err == nil {
+		id = value
+		userRequestedVnid = true
+		netid.DeleteWantsVNID(ns)
+	} else if err != netid.ErrorVNIDNotFound {
+		return err
+	} else if _, err = netid.GetVNID(ns); err == nil {
+		// Nothing to do in this case
+		// VNID already allocated and namespace does not want new VNID
+		return nil
+	}
+
+	name := ns.ObjectMeta.Name
+	if userRequestedVnid {
+		// Only accept global VNID or VNIDs that are already allocated
+		if (id != netid.GlobalVNID) && !c.vnid.Has(id) {
+			return fmt.Errorf("requested netid %d not allocated", id)
+		}
+	} else {
+		// Do vnid allocation
+		if c.globalNamespaces.Has(name) {
+			id = netid.GlobalVNID
+		} else {
+			id, err = c.vnid.AllocateNext()
+			if err != nil {
+				return err
+			}
+			tx.Add(func() error { return c.vnid.Release(id) })
+		}
+	}
+	if err := netid.SetVNID(ns, id); err != nil {
+		return err
+	}
+
+	// TODO: could use a client.GuaranteedUpdate/Merge function
+	for i := 0; i < retryCount; i++ {
+		_, err := c.client.Update(ns)
+		if err == nil {
+			if userRequestedVnid {
+				glog.Infof("Updated netid %d for namespace %q", id, name)
+			} else {
+				glog.Infof("Assigned netid %d for namespace %q", id, name)
+			}
+			// commit and exit
+			tx.Commit()
+			return nil
+		}
+
+		if errors.IsNotFound(err) {
+			return nil
```

Comment:
When we try to get/update the namespace and if the namespace got deleted (async by some other process), then IsNotFound(err) returns true. In this case, we don't have to patch anything so we don't return an error.